config-service-impl/src/main/java/org/hypertrace/config/service/store/DocumentConfigStore.java

```diff
@@ -208,30 +212,89 @@ public Map<ConfigResourceContext, ContextSpecificConfig> getContextConfigs(
-        ConfigDocument configDocument = ConfigDocument.fromJson(documentString);
-        String context = configDocument.getContext();
-        if (seenContexts.add(context)) {
-          convertToContextSpecificConfig(configDocument).ifPresent(contextSpecificConfigList::add);
-        }
+        processDocument(documentIterator.next(), seenContexts, configList);
       }
     }
-    Collections.sort(
-        contextSpecificConfigList,
+
+    configList.sort(
         Comparator.comparingLong(ContextSpecificConfig::getCreationTimestamp).reversed());
-    return contextSpecificConfigList;
+    return configList;
+  }
+
+  private Query buildQuery(
+      ConfigResource configResource,
+      org.hypertrace.config.service.v1.Filter filter,
+      org.hypertrace.config.service.v1.Pagination pagination,
+      List<SortBy> sortByList) {
+
+    FilterTypeExpression combinedFilter = getCombinedFilter(configResource, filter);
+    Query.QueryBuilder queryBuilder = Query.builder().setFilter(combinedFilter);
+    if (pagination != null
+        && !pagination.equals(org.hypertrace.config.service.v1.Pagination.getDefaultInstance())) {
+      queryBuilder.setPagination(
+          Pagination.builder().offset(pagination.getOffset()).limit(pagination.getLimit()).build());
+    }
+
+    if (sortByList != null) {
+      sortByList.forEach(
+          sortBy ->
+              queryBuilder.addSort(
+                  IdentifierExpression.of(sortBy.getSelection().getConfigJsonPath()),
+                  convertSortOrder(sortBy)));
+    }
+
+    queryBuilder.addSort(IdentifierExpression.of(VERSION_FIELD_NAME), SortOrder.DESC);
+    return queryBuilder.build();
+  }
+
+  private FilterTypeExpression getCombinedFilter(
+      ConfigResource configResource, org.hypertrace.config.service.v1.Filter additionalFilter) {
+
+    FilterTypeExpression resourceFilter = getConfigResourceFilterTypeExpression(configResource);
+    if (additionalFilter == null
+        || additionalFilter.equals(org.hypertrace.config.service.v1.Filter.getDefaultInstance())) {
+      return resourceFilter;
+    }
+
+    FilterTypeExpression docStoreFilter =
+        filterExpressionBuilder.buildFilterTypeExpression(additionalFilter);
+    return LogicalExpression.and(resourceFilter, docStoreFilter);
+  }
+
+  @SneakyThrows
+  private void processDocument(
+      Document document, Set<String> seenContexts, List<ContextSpecificConfig> configList) {
+
+    ConfigDocument configDocument = ConfigDocument.fromJson(document.toJson());
+    String context = configDocument.getContext();
+
+    if (seenContexts.add(context)) {
+      convertToContextSpecificConfig(configDocument).ifPresent(configList::add);
+    }
+  }
+
+  private static SortOrder convertSortOrder(SortBy sortBy) {
+    switch (sortBy.getSortOrder()) {
+      case SORT_ORDER_DESC:
+        return SortOrder.DESC;
+      case SORT_ORDER_ASC:
+      default:
+        return SortOrder.ASC;
```

Comment:
Currently updated ASC as default, please let me know if these needs to be changed

Reply:
That's fine as long as it only applies if a sort field is specified but no direction (which I believe is the case here)

Reply:
Yes